### PR TITLE
Fix operator precedence in the parser

### LINKED ISSUE #1457: More operator precedence issues with `dhall format`
These are similar to #1455 but AFAICT couldn't arise with well-typed expressions. Nevertheless they should be fixed, if only so the idempotence test can proceed further to more interesting issues.

#### `||` and `+`

```
*** Failed! Falsifiable (after 314 tests and 3 shrinks):
Unicode
(Header "",BoolOr (NaturalLit 0) (NaturalPlus (NaturalLit 0) NaturalEven))
"0 || 0 + Natural/even\n" /= "(0 || 0) + Natural/even\n"
Use --quickcheck-replay=541381 to reproduce.
```

```
$ echo 'a || b + c' | dhall format
(a || b) + c
```
The added parentheses are wrong, This should be parsed as `a || (b + c)`.

#### `//` and `==`

```
*** Failed! Falsifiable (after 55 tests and 2 shrinks):
Unicode
(Header "",Prefer None (BoolEQ (Const Sort) (DoubleLit (DhallDouble {getDhallDouble = 0.0}))))
"None \11005 Sort == 0.0\n" /= "(None \11005 Sort) == 0.0\n"
Use --quickcheck-replay=411136 to reproduce.
```
```
$ echo 'a // b == c' | dhall format | dhall format
(a ⫽ b) == c
```

Again this should be parsed as `a // (b == c)`

#### `//` and `//\\`

```
*** Failed! Falsifiable (after 113 tests and 5 shrinks):
ASCII
(Header "{-,an1P\1035889O\t-}\n{-1\659625-}\n",Prefer (NaturalLit 0) (CombineTypes (Var (V "a" 0)) (BoolLit False)))
"{-,an1P\1035889O\t-}\n{-1\659625-}\n0 // a //\\\\ False\n" /= "{-,an1P\1035889O\t-}\n{-1\659625-}\n(0 // a) //\\\\ False\n"
Use --quickcheck-replay=938268 to reproduce.
```

As above.

#### `++` and `#`

```
*** Failed! Falsifiable (after 758 tests and 3 shrinks):
Unicode
(Header "",TextAppend ListLength (ListAppend (Const Type) NaturalBuild))
"List/length ++ Type # Natural/build\n" /= "(List/length ++ Type) # Natural/build\n"
Use --quickcheck-replay=960143 to reproduce.
```

#### `&&` and `/\`

```
*** Failed! Falsifiable (after 758 tests and 3 shrinks):
Unicode
(Header "",TextAppend ListLength (ListAppend (Const Type) NaturalBuild))
"List/length ++ Type # Natural/build\n" /= "(List/length ++ Type) # Natural/build\n"
Use --quickcheck-replay=960143 to reproduce.
```

If we can distill well-typed test cases from these, we should add them to the `dhall-lang` acceptance tests.

### PR DESCRIPTION
**Pull request by sjakobi**
Previously some operator expressions were associated incorrectly, e.g.

* a || b + c  -> (a || b) + c
* a // b == c -> (a // b) == c

Fixes #1457.

### CODE REVIEW

File: dhall/src/Dhall/Parser/Expression.hs [hunk 1]
```diff
@@ -257,9 +257,10 @@ parsers embedded = Parsers {..}
 
             alternative4A <|> alternative4B <|> pure a
 
-    operatorExpression = precedence0Expression
+    operatorExpression =
+        foldr makeOperatorExpression applicationExpression operatorParsers
```
Comment (sjakobi) [inline dhall/src/Dhall/Parser/Expression.hs]: `foldr` is just damn powerful! :muscle:

**Comment by sjakobi** [comment]
This does seem to impact parser performance though: The `Issue #108/Text` benchmark now takes ~7.8ms instead of ~6.9ms. Not sure whether that's acceptable…

We could consider putting this code behind a feature flag and mostly use it to discover more interesting formatting idempotence issues.

**Comment by sjakobi** [comment]
@Gabriel439 Your comment https://github.com/dhall-lang/dhall-haskell/pull/1460#issuecomment-545256369 seems to apply this PR:

> Yeah, the reason it's grouped like that is a performance optimization since you only run into expressions with the wrong precedence if they are ill-typed expressions

What do you think about the idea of putting the this code behind a cabal flag? The idempotence test is fairly useless as long as it can stumble over these operator precedence issues.

**Comment by Gabriel439** [comment]
@sjakobi: The benchmark I'm most interested in is this one:

https://github.com/dhall-lang/dhall-haskell/blob/16e6a12c3da1966b095cc8089d8c1274c8fca4e5/dhall/benchmark/parser/Main.hs#L98

That's the one where I'd expect this optimization to have the largest impact.  What is the difference that you observe for that benchmark?

**Comment by sjakobi** [comment]
Previously 346 μs, now ~442 μs. I guess that's a bit much! :/

**Comment by sjakobi** [comment]
I've tried to measure the impact on real code using `cpkg`'s `pkgs-set.dhall`:

#### `master`

```
$ bench "dhall resolve --no-cache --file pkgs/pkg-set.dhall  --immediate-dependencies > /dev/null"
benchmarking dhall resolve --no-cache --file pkgs/pkg-set.dhall  --immediate-dependencies > /dev/null
time                 173.0 ms   (172.1 ms .. 173.7 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 172.6 ms   (172.1 ms .. 172.9 ms)
std dev              498.9 μs   (143.1 μs .. 734.1 μs)
variance introduced by outliers: 12% (moderately inflated)
```

#### This branch

```
$ bench "dhall resolve --no-cache --file pkgs/pkg-set.dhall  --immediate-dependencies > /dev/null"
benchmarking dhall resolve --no-cache --file pkgs/pkg-set.dhall  --immediate-dependencies > /dev/null
time                 187.6 ms   (185.5 ms .. 189.0 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 188.5 ms   (187.9 ms .. 189.3 ms)
std dev              955.5 μs   (721.7 μs .. 1.153 ms)
variance introduced by outliers: 14% (moderately inflated)
```

Is there a better command for measuring parser performance?

**Comment by Gabriel439** [comment]
@sjakobi: Actually, I think that's fine.  Sure it's a measurable performance regression but it's not that bad and now it's more correct.  After all, the contract of `dhall format` is that it should work even for ill-typed code, so if it misformats such code then it's not working as advertised.  I just wanted to make sure that we measured the performance regression; that's all.

**Pull request merged**